common/logger/derive/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use proc_macro::TokenStream;
+use proc_macro2::Ident;
+use quote::{format_ident, quote};
+use syn::{
+    parse_macro_input, AngleBracketedGenericArguments, Attribute, Data, DataStruct, DeriveInput,
+    Fields, FieldsNamed, GenericArgument, Meta, MetaList, NestedMeta, Path, PathArguments,
+    PathSegment, Type, TypePath,
+};
+
+#[proc_macro_derive(Schema, attributes(schema))]
+pub fn derive(input: TokenStream) -> TokenStream {
+    // Parse the input tokens into a syntax tree.
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = input.ident;
+    let fields = match input.data {
+        Data::Struct(DataStruct {
+            fields: Fields::Named(FieldsNamed { named, .. }),
+            ..
+        }) => named,
+        _ => panic!("derive(Schema) only supports structs with named fields"),
+    };
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
+    let fields: Vec<StructField> = fields
+        .iter()
+        .map(|f| {
+            let ty = f.ty.clone();
+            let value_type = extract_attr(&f.attrs);
+
+            let inner_ty = extract_internal_type(&ty).cloned();
+            StructField {
+                value_type,
+                ident: f.ident.clone(),
+                ty: f.ty.clone(),
+                inner_ty,
+            }
+        })
+        .collect();
+
+    let setters = fields.iter().map(|f| {
+        let ident = &f.ident;
+
+        if let Some(ty) = &f.inner_ty {
+            quote! {
+                pub fn #ident(mut self, #ident: #ty) -> Self {
+                    self.#ident = ::std::option::Option::Some(#ident);
+                    self
+                }
+            }
+        } else {
+            let ty = &f.ty;
+            quote! {
+                pub fn #ident(mut self, #ident: #ty) -> Self {
+                    self.#ident = #ident;
+                    self
+                }
+            }
+        }
+    });
+
+    // Calls to visit_pair
+    let visitor = format_ident!("visitor");
+    let from_serde = quote! { ::libra_logger::Value::from_serde };
+    let from_display = quote! { ::libra_logger::Value::from_display };
+    let from_debug = quote! { ::libra_logger::Value::from_debug };
+    let key_new = quote! { ::libra_logger::Key::new };
+    let visits = fields.iter().map(|f| {
+        let ident = f.ident.as_ref().unwrap();
+        let ident_str = ident.to_string();
+
+        let from_fn = match f.value_type {
+            Some(ValueType::Display) => &from_display,
+            Some(ValueType::Debug) => &from_debug,
+            Some(ValueType::Serde) | None => &from_serde,
```

Comment:
is this documented somewhere that the default is from serde?

Reply:
I'll make sure to document that in the example I added to the docs